src/provider/pedersen.rs

```diff
@@ -167,15 +168,51 @@ where
   type Commitment = Commitment<E>;
 
   fn setup(label: &'static [u8], n: usize) -> Self::CommitmentKey {
+    let gens = E::GE::from_label(label, n.next_power_of_two() + 1);
+
+    let (h, ck) = gens.split_first().unwrap();
+
     Self::CommitmentKey {
-      ck: E::GE::from_label(label, n.next_power_of_two()),
+      ck: ck.to_vec(),
+      h: Some(h.clone()),
     }
   }
 
-  fn commit(ck: &Self::CommitmentKey, v: &[E::Scalar]) -> Self::Commitment {
+  fn commit(ck: &Self::CommitmentKey, v: &[E::Scalar], r: &E::Scalar) -> Self::Commitment {
     assert!(ck.ck.len() >= v.len());
+
+    if ck.h.is_some() {
+      let mut scalars: Vec<E::Scalar> = v.to_vec();
+      scalars.push(*r);
+      let mut bases = ck.ck[..v.len()].to_vec();
+      bases.push(ck.h.as_ref().unwrap().clone());
+
+      Commitment {
+        comm: E::GE::vartime_multiscalar_mul(&scalars, &bases),
+      }
+    } else {
+      assert_eq!(*r, E::Scalar::ZERO);
+
+      Commitment {
+        comm: E::GE::vartime_multiscalar_mul(v, &ck.ck[..v.len()]),
+      }
+    }
+  }
+
+  fn derandomize(
```

Comment:
I'm not sure this method needs to be this complicated and involve MSMs. Why not do something like:

`fn derandomize(dk: &Self::DerandKey, C: &Self::Commitment, blind: E::Scalar) -> Self::Commitment {
    Commitment {
      comm: C.comm - <E::GE as DlogGroup>::group(&dk.h) * blind,
    }
  }`